src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -583,6 +576,252 @@ private bool TryRemoveInvocation(string invocationId, out InvocationRequest irq)
+
+                    try
+                    {
+                        if (buffer.IsEmpty && result.IsCompleted)
+                        {
+                            // The server disconnected
+                            break;
+                        }
+                        else if (result.IsCanceled)
+                        {
+                            // We aborted because the server timed-out
+                            closeException = new TimeoutException(
+                                $"Server timeout ({ServerTimeout.TotalMilliseconds:0.00}ms) elapsed without receiving a message from the server.");
+                            break;
+                        }
+
+                        if (timeoutTimer != null)
+                        {
+                            Log.ResettingKeepAliveTimer(_logger);
+                            timeoutTimer.Change(ServerTimeout, Timeout.InfiniteTimeSpan);
+                        }
+
+                        // We have data, process it
+                        var (close, exception) = await ProcessMessagesAsync(buffer);
+                        if (close)
+                        {
+                            // Closing because we got a close frame, possibly with an error in it.
+                            closeException = exception;
+                            break;
+                        }
+                    }
+                    finally
+                    {
+                        _connection.Transport.Input.AdvanceTo(consumed, examined);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Log.ServerDisconnectedWithError(_logger, ex);
+                closeException = ex;
+            }
+            finally
+            {
+                // Dispose the connection
+                await _connection.DisposeAsync();
+                _connection = null;
+
+                // Stop the timeout timer.
+                timeoutTimer?.Dispose();
+            }
+
+            if (closeException != null)
+            {
+                Log.ShutdownWithError(_logger, closeException);
+            }
+            else
+            {
+                Log.ShutdownConnection(_logger);
+            }
+
+            // Cancel any outstanding invocations
+            CancelOutstandingInvocations(closeException);
+
+            // Fire-and-forget the closed event
+            _ = Task.Run(() =>
+            {
+                try
+                {
+                    Log.InvokingClosedEventHandler(_logger);
+                    Closed?.Invoke(closeException);
+                }
+                catch (Exception ex)
+                {
+                    Log.ErrorDuringClosedEvent(_logger, ex);
+                }
+            });
+        }
+
+        private ValueTask<FlushResult> WriteAsync(byte[] payload) =>
+            _connection.Transport.Output.WriteAsync(payload);
+
```

Comment:
Did you introduce a lock around the write?

Reply:
It's always done inside the `_startLock`

Reply:
I hate that the start lock isn't just for starting? Can you rename it?

Reply:
Oh yeah, it's a terrible name. Will rename.